feat: introducing "copy_terraform_lock_file" to fine tune Lock File Handling

rodrigorfk opened the pull request:
<!-- Prepend '[WIP]' to the title if this PR is still a work-in-progress. Remove it when it is ready for review! -->

## Description

Fixes #1653.

This PR introduces a new attribute called `copy_terraform_lock_file` under the `terraform` block as following:

```hcl
terraform {
  ...
  copy_terraform_lock_file = false
}
```

The default value is `true` and does not change any existing Terragrunt behaviour, once it is defined as `false`, Terragrunt will stop copying the `.terraform.lock.hcl` file from the temporary terragrunt directory to the working directory.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [ ] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

<!-- One-line description of the PR that can be included in the final release notes. -->
Added / Removed / Updated [X].

Added "copy_terraform_lock_file" attribute to fine tune Lock File Handling

### Migration Guide

<!-- Important: If you made any backward incompatible changes, then you must write a migration guide! -->

rodrigorfk commented:
Hi @denis256, would you be so kind and provide some feedback about this PR? Thanks in advance.

denis256 commented:
Hi, yes, it is in my queue to check... :hourglass:

rodrigorfk commented:
Hi @denis256 , sorry to bother you once again, but is there anything I could help you to expedite the review of this PR? I would love to hear from you if this change is something the project is willing to accept. Thanks once again.

Josephuss commented:
hello @rodrigorfk @denis256 , hope you're doing well. Can I ask if this merge can be completed sometime soon? 

my team found this merge request and  we could all use this feature since we're trying to create a remote and client setup for terragrunt and versioning could become an issue if there are locks on both remote and client side.

rodrigorfk commented:
Hi @Josephuss , thanks for your message, I'm happy to see that your team also thought this change would be useful, I have updated the branch and fixed the outstanding conflicts.

james-masson commented:
We've also come across this PR, it's exactly the functionality we need, for the same reasons as the others mentioned.

IvanKuzyshyn commented:
Thanks @rodrigorfk for bringing this update! This change unlocks the ability for the lock file to be managed and versioned by the remove module itself. It would be great to have it merged.

rodrigorfk commented:
Thanks for the interest @IvanKuzyshyn , I have updated the branch and resolved the conflicts once again. Hopefully @denis256 will have some time to provide a feedback about this change soon. 🙏

gmauleon commented:
Just passing by to nudge this as I'm also interested, thanks in advance 😀

yhakbar commented on docs/_docs/02_features/lock-file-handling.md:
```diff
@@ -96,3 +96,16 @@ should end up looking something like this:
+### Disabling the copy of the generated lock file
+
+In certain use cases, like when using a remote source module, containing the lock file within it, you probability
+don't want Terragrunt to also copy the lock file into you working directory, in such scenarios you can opt out from the copy
```
```suggestion
don't want Terragrunt to also copy the lock file into your working directory. In these scenarios you can opt-out of copying
```

yhakbar commented on docs/_docs/02_features/lock-file-handling.md:
```diff
@@ -96,3 +96,16 @@ should end up looking something like this:
+
+In certain use cases, like when using a remote source module, containing the lock file within it, you probability
+don't want Terragrunt to also copy the lock file into you working directory, in such scenarios you can opt out from the copy
+feature by using `copy_terraform_lock_file = false` in the `terragrunt.hcl` file as following:
```
```suggestion
the `.terraform.lock.hcl` file by using `copy_terraform_lock_file = false` in the `terraform` configuration block as follows:
```

yhakbar commented on docs/_docs/04_reference/config-blocks-and-attributes.md:
```diff
@@ -104,6 +104,11 @@ The `terraform` block supports the following arguments:
 
+- `copy_terraform_lock_file` (attribute): In certain use cases, you don't want to check the terraform provider lock
+  file into your source repository from your working directory as described in
+  [Lock File Handling]({{site.baseurl}}/docs/features/lock-file-handling/). This attribute allows you disabling the copy
```
```suggestion
  [Lock File Handling]({{site.baseurl}}/docs/features/lock-file-handling/). This attribute allows you to disable the copy
```

yhakbar reviewed (changes requested):
This looks great! I left some small details that can be addressed before we get this merged in.

Sorry we've let this PR get stale. I appreciate that the tests were updated, and docs were improved with the feature update.

rodrigorfk commented:
Hi @yhakbar, thanks for the review and suggestions, I will be away for a week and will make sure I go over your comments once I am back.

rodrigorfk commented on cli/commands/terraform/action.go:
```diff
@@ -347,7 +347,11 @@ func confirmActionWithDependentModules(ctx context.Context, terragruntOptions *o
-func shouldCopyLockFile(args []string) bool {
+func shouldCopyLockFile(args []string, terraformConfig *config.TerraformConfig) bool {
+	if terraformConfig != nil && terraformConfig.CopyTerraformLockFile != nil && !*terraformConfig.CopyTerraformLockFile {
+		return false
```
thanks for the suggestion @yhakbar, but I am not sure if we should perform an early return like that, if we do that and the user explicitly defines `copyTerraformLockFile = true` on their terragrunt.hcl, terragrunt will copy the file on all commands, which will change the existing behaviour about copying it only for the `init` and `providers lock` command. Thoughts?

yhakbar replied:
Ah, good point. It might help to leave a comment above the check to make sure that someone doesn't make the same logical mistake I'm making.

rodrigorfk replied:
done @yhakbar, added the comment, do you mind having another look? thanks in advance

gmauleon commented:
Hey there kind folks, just coming to bump that up the grave if anyone is available for review 🙏

yhakbar commented:
Hey @gmauleon , this PR currently has some conflicts, so it'll need those conflicts resolved, lints passing and tests run before it can be reviewed. When it is, the title should be updated, with reviews requested.

If you could help out @rodrigorfk by supplying a PR to his source branch, that might help things move along.

gmauleon commented:
> Hey @gmauleon , this PR currently has some conflicts, so it'll need those conflicts resolved, lints passing and tests run before it can be reviewed. When it is, the title should be updated, with reviews requested.
> 
> If you could help out @rodrigorfk by supplying a PR to his source branch, that might help things move along.

I'm willing to try, I got a local branch rebased and I'm fixing tests atm, how would I proceed to update this PR afterwards?

gmauleon commented:
Yeah not sure what to do, I did a rebase of the official main, so I got a bunch of conflict with the original PR branch because a force push would be needed. Branch is here: https://github.com/gmauleon/terragrunt/tree/fix-feature/copy-lock-file

Let me know what I can do to help...

gmauleon commented:
In case it can make things easier for everyone, I created a PR based on this one at https://github.com/gruntwork-io/terragrunt/pull/3429.

gmauleon commented on config/include.go:
```diff
@@ -278,6 +278,12 @@ func (cfg *TerragruntConfig) Merge(sourceConfig *TerragruntConfig, terragruntOpt
+				cfg.Terraform.CopyTerraformLockFile = sourceConfig.Terraform.CopyTerraformLockFile
+			}
+
+			mergeExtraArgs(terragruntOptions, sourceConfig.Terraform.ExtraArgs, &cfg.Terraform.ExtraArgs)
```
I believe there is an extra call to `mergeExtraArgs` here

rodrigorfk replied:
you are right, it was indeed duplicated, thanks for catching this up.

gmauleon commented on docs/_docs/02_features/lock-file-handling.md:
```diff
@@ -96,3 +96,16 @@ should end up looking something like this:
+### Disabling the copy of the generated lock file
+
+In certain use cases, like when using a remote module containing a lock file within it, you probably
+don't want Terragrunt to also copy the lock file into your working directory. In these scenarios you can opt-out of copying
```
```suggestion
don't want Terragrunt to also copy the lock file into your working directory. In these scenarios, you can opt-out of copying
```

gmauleon commented:
> Hi @gmauleon, sorry for the late response, I've updated the feature branch once again based on the latest changes from `main`, could you please take a look and check if this is what you were looking for? thanks in advance.

Hey @rodrigorfk thanks for this! I did a couple of small remarks myself, wdyt?
I believe you could then remove the WIP from the title and request  a real review from the code owners 😁

rodrigorfk commented:
@gmauleon, sounds good to me, I've applied your suggestion and fixed the double function call you caught. The `WIP` was removed from the title and the PR is once again ready for review.

gmauleon commented:
Cool, thanks again. We'll let @yhakbar look it up when he has time. I'll close my now useless duplicate [PR](https://github.com/gruntwork-io/terragrunt/pull/3429).

Thanks again!

denis256 commented:
It looks like TestRenderJsonMetadataTerraform still fails

![image](https://github.com/user-attachments/assets/7ec115e0-23e8-4b7a-89d0-65bba1033ac2)

gmauleon commented:
Ah indeed, tests are taking ages on my machine and I don't have access to some AWS credentials, so I just ran specific tests 🙄
Hopefully it's the only remaining one, @rodrigorfk  this [PR](https://github.com/rodrigorfk/terragrunt/pull/2/files) but if you can run all tests on your side it's even better.

Sorry for all the spam guys, thanks in advance.

rodrigorfk commented:
Thanks for the feedback @denis256 , the `TestRenderJsonMetadataTerraform` test was fixed and it is supposed to be working as expected now.

yhakbar commented:
It looks like tests are passing now, and I don't see anything obviously wrong with this PR.

To keep things moving, I'm going to merge this PR, then sync-up with @denis256 to see if any other changes should be introduced before releasing this new feature.
